### PR TITLE
Store config description in Airflow configuration object

### PR DESCRIPTION
**Pull request by potiuk**
We would like to use the config.yml approach as our main source of truth
for airflow configuration. So far it has been split amongst multiple
files:

  * confg.yml -> descroption
  * default_airflow.cfg -> JINJA template to keep cofiguration, examples
    and description (used to generate airflow.cfg when airflow starts)
  * default_test.cfg -> storing test-only configuration used in some
    test cases - used to generate unittest.cfg
  * scripts/in_container/airflow_ci.cfg where dCI-specific configuration
    overwrote the unittest.cfg

This change consolidates it all into unified appraoch, where all
configuration information is retrieved from .yml files stored in
config_templates. No more additional template files processed by
JINJA, no more special CI versions of it, no more unittestdb.cfg file
where such configuration would be generated, no more unittestdb to
be used separately for tests.

* The default_*.cfg files were not real configuration files, becuase
  they were really JINJA templates and it got peoeple confused when
  copying the files. This change leaves the file empty with the
  comment that instructs the user how they can get the default
  configuration.
* The default_airflow.cfg is gone and instead, we have a way to
  show defaults via `airflow config list --defaults` command
* Unittest config is generated on-the-flight using defaults stored
  in confg_templates. constituing a single place where they need
  to be changed for the tests to use them
* internally, description of the configuration is stored in
  AirflowConfigurationParser and used by the parser to generate
  the default configuration when needed.
* we got rid of `{{{{` in templated config defaults by marking
  the templates with ``is_template`` and getting rid of processing
  those entries with regular formatting when generating the default
  values. This only concerns defaults from config.yml. Reading
  those configuration entries from file is unaffected.

This change aims to be 100% backwards compatible with the previous
implementation when it comes to functionality, even if internals
changed. It also does not add provider-specific changes that are
coming separately.

The only changes visible to the user are:

* generated airflow.cfg is slightly more
  readable and displays names of variables that can be used to override
  each configuration (which is very useful for copy&pasting)

* user are advised, instead of copying the default_airflow.cfg to use
  `airflow config list --defaults` to generate production config. This
  configuration has all the entries commented out, so that they can
  selectively uncomment and change the values they want. This is now
  promoted as "best practice" in the documentation.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by potiuk** [comment]
This is the most important part extracted from #32604 (I will shortly rebase #32604) on top of this one. This one DOES NOT contain any provider-related changes. Those will come as finall installment of the provider config contributed changes and is a crucial refactor to be able to add provider-contributed configurations. 

I rewrote a little what I have done initially after I run more thorough testing - I came back (and got friends with) using ConfigParser as "default" configuration storage (because of interpolation). I also think that having `unit_test.cfg` as source of defaults for tests is better and more readable than .yml as it was in the original PR, so I changed it back in this one.

The gist of this change remains the same as described in the original PR - make `config.yaml` the ONLY source of truth for configuration defaults, keep test config entirely in the sources of airflow - and do not genarate the unittest.cfg file on the filesystem but use test configuration from memory - thus making our tests for configuraiton repeatable and rerunnable, no matter what environment they are running it (breeze, local venv, pycharm). 

I also included some comments from the review in #32604 by @jedcunningham @ferruzzi and @o-nikolas (some of them still will need to be addressed separately and I resolved all those that are applied).

File: TESTING.rst [hunk 1]
```diff
@@ -55,8 +55,36 @@ Follow the guidelines when writing unit tests:
...
+
+Some of the unit tests require special configuration set as the ``default``. This is done automatically by
+adding ``AIRFLOW__CORE__UNIT_TEST_MODE=True`` to the environment variables in Pytest auto-used
+fixture. This in turn makes airflow configuration load test configuration from
+``airflow/config_templates/unit_tests.cfg`` - the test configuration from there replaces the original
+defaults from ``airflow/config_templates/config.yml``. If you want to add some test-only configuration,
+as default for all tests you should add the value to this file.
+
+You can also of course override the values in individual test by patching environment variables following
+the usual ``AIRFLOW__SECTION__KEY`` pattern.
```
Comment (potiuk) [inline TESTING.rst]:
```suggestion
the usual ``AIRFLOW__SECTION__KEY`` pattern or using ``conf_vars`` context manager.
```

File: airflow/cli/cli_config.py [hunk 2]
```diff
@@ -90,7 +91,7 @@ def _check_value(self, action, value):
                         f"executors derived from them, your current executor: {executor}, subclassed from: "
                         f'{", ".join([base_cls.__qualname__ for base_cls in executor_cls.__bases__])}'
                     )
-                    raise ArgumentError(action, message)
+                    warnings.warn(message)
```
Comment (Lee-W) [inline airflow/cli/cli_config.py]: I am wondering why this is changed from exception to warning.

File: airflow/config_templates/config.yml [hunk 3]
```diff
@@ -2931,7 +2935,8 @@ kubernetes_executor:
     tcp_keep_cnt:
       description: |
         When the `enable_tcp_keepalive` option is enabled, if Kubernetes API does not respond
-        to a keepalive probe, TCP retransmits the probe `tcp_keep_cnt number` of times before
+        to a keepalive probe, TCP retransmits the probe `tcp_keep_cnt number` o
+        f times before
```
Comment (Lee-W) [inline airflow/config_templates/config.yml]:
```suggestion
        to a keepalive probe, TCP retransmits the probe `tcp_keep_cnt number` of times before
```

File: airflow/config_templates/config.yml.schema.json [hunk 4]
```diff
@@ -72,6 +72,11 @@
                 "sensitive": {
                     "type": "boolean",
                     "description": "When true, this option is sensitive and can be specified using AIRFLOW__{section}___{name}__SECRET or AIRFLOW__{section}___{name}_CMD environment variables. See: airflow.configuration.AirflowConfigParser.sensitive_config_values"
+                },
+                "is_template": {
+                    "type": "boolean",
+                    "description": "When true, this option is not processed by format after being processed by AirflowConfigParser, The JINJA templates should be encoded as {{{{ }}}} (they are converted to {{ }} in the final field default), whereas the Python format strings should be encoded as `{{ }}` and they are converted to { } in the final field default)",
+                    "default": false
```
Comment (amoghrajesh) [inline airflow/config_templates/config.yml.schema.json]:
```suggestion
                    "description": "When true, this option is not processed by formatter after being processed by AirflowConfigParser, The JINJA templates should be encoded as {{{{ }}}} (they are converted to {{ }} in the final field default), whereas the Python format strings should be encoded as `{{ }}` and they are converted to { } in the final fields)",
                    "default": false
```

File: airflow/config_templates/default_airflow.cfg [hunk 5]
```diff
@@ -16,1507 +15,19 @@
...
-# state in the metadata database.
-parallelism = 32
-
-# The maximum number of task instances allowed to run concurrently in each DAG. To calculate
-# the number of tasks that is running concurrently for a DAG, add up the number of running
-# tasks for all DAG runs of the DAG. This is configurable at the DAG level with ``max_active_tasks``,
-# which is defaulted as ``max_active_tasks_per_dag``.
+# This file used to have something that was similar to the default Airflow configuration but it was
+# really just a template. It was used to generate the final configuration and it was confusing
+# if you copied it to your configuration and some of values were wrong.
```
Comment (amoghrajesh) [inline airflow/config_templates/default_airflow.cfg:25]:
```suggestion
NOTE
# This file used to have something that was similar to the default Airflow configuration but it was
# really just a template. It was used to generate the final configuration and it was confusing
# if you copied it to your configuration and some of values were wrong.
```

**Review by amoghrajesh — COMMENTED** [review]
Did one pass at the PR and have few nits. Nothing major

File: TESTING.rst [hunk 6]
```diff
@@ -55,8 +55,36 @@ Follow the guidelines when writing unit tests:
   tests, so we run Pytest with ``--disable-warnings`` but instead we have ``pytest-capture-warnings`` plugin that
   overrides ``recwarn`` fixture behaviour.
 
-**NOTE:** We plan to convert all unit tests to standard "asserts" semi-automatically, but this will be done later
-in Airflow 2.0 development phase. That will include setUp/tearDown/context managers and decorators.
+
+.. note::
+
+  We are in the process of convert all unit tests to standard "asserts" and pytest fixtures
```
Comment (phanikumv) [inline TESTING.rst]:
```suggestion
  We are in the process of converting all unit tests to standard "asserts" and pytest fixtures
```

File: TESTING.rst [hunk 7]
```diff
@@ -55,8 +55,36 @@ Follow the guidelines when writing unit tests:
...
+  so if you find some tests that are still using classic setUp/tearDown approach or unittest asserts, feel
+  free to convert them to pytest.
+
+Airflow configuration for unit tests
+------------------------------------
+
+Some of the unit tests require special configuration set as the ``default``. This is done automatically by
+adding ``AIRFLOW__CORE__UNIT_TEST_MODE=True`` to the environment variables in Pytest auto-used
+fixture. This in turn makes airflow configuration load test configuration from
```
Comment (phanikumv) [inline TESTING.rst]:
```suggestion
fixture. This in turn makes airflow load test configuration from
```

File: TESTING.rst [hunk 8]
```diff
@@ -55,8 +55,36 @@ Follow the guidelines when writing unit tests:
...
+defaults from ``airflow/config_templates/config.yml``. If you want to add some test-only configuration,
+as default for all tests you should add the value to this file.
+
+You can also of course override the values in individual test by patching environment variables following
+the usual ``AIRFLOW__SECTION__KEY`` pattern or ``conf_vars`` context manager.
+
+.. note::
+
+  The test configuration for Airflow before July 2023 was automatically generated in a file named
+  ``AIRFLOW_HOME/unittest.cfg`` and template for it was stored in "config_templates" next to the yaml file,
```
Comment (phanikumv) [inline TESTING.rst]:
```suggestion
  ``AIRFLOW_HOME/unittest.cfg``.  The template for it was stored in "config_templates" next to the yaml file,
```

File: TESTING.rst [hunk 9]
```diff
@@ -55,8 +55,36 @@ Follow the guidelines when writing unit tests:
...
+  however this was only done for the first time you run airflow and you had to manually maintain the file,
```
Comment (phanikumv) [inline TESTING.rst]:
```suggestion
  however this was only done for the first time you run airflow and you had to manually maintain the file.
```

File: TESTING.rst [hunk 10]
```diff
@@ -55,8 +55,36 @@ Follow the guidelines when writing unit tests:
...
+  and it was pretty arcane knowledge, also this file has been overwritten in Breeze environment
```
Comment (phanikumv) [inline TESTING.rst]: Can you please mention which file has been overwritten for more clarity.

Comment (phanikumv) [inline TESTING.rst]:
```suggestion
 It was pretty arcane knowledge, and this file has been overwritten in Breeze environment
```

File: airflow/cli/cli_config.py (hunk 2 above)
Comment (potiuk) [reply airflow/cli/cli_config.py]: Good point. I used it for testing of celery and it remained here. I was going to propose this as a change but indeed it is not needed in this PR. will remove it.

File: airflow/config_templates/config.yml (hunk 3 above)
Comment (potiuk) [reply airflow/config_templates/config.yml]: Right.

File: airflow/config_templates/config.yml.schema.json (hunk 4 above)
Comment (potiuk) [reply airflow/config_templates/config.yml.schema.json]: Actually this description is slightly wrong (I managed to change it in the way that we do not need `{{{{` any more. I will update the description to be way shorter as well.

File: airflow/config_templates/default_airflow.cfg (hunk 5 above)
Comment (potiuk) [reply airflow/config_templates/default_airflow.cfg:25]: reworded slightly.

File: TESTING.rst (hunk 6 above)
Comment (potiuk) [reply TESTING.rst]: Yeah.

File: TESTING.rst (hunk 7 above)
Comment (potiuk) [reply TESTING.rst]: Reworded this paragraph and made sentences shorter.

File: TESTING.rst (hunk 8 above)
Comment (potiuk) [reply TESTING.rst]: Yeah

File: TESTING.rst (hunk 10 above)
Comment (potiuk) [reply TESTING.rst]: Yeah. good point.

File: TESTING.rst [hunk 11]
```diff
@@ -55,8 +55,37 @@ Follow the guidelines when writing unit tests:
...
+
+.. note::
+
+  The test configuration for Airflow before July 2023 was automatically generated in a file named
+  ``AIRFLOW_HOME/unittest.cfg``. The template for it was stored in "config_templates" next to the yaml file.
+  However writing the file was only done for the first time you run airflow and you had to manually
+  maintain the file. It was pretty arcane knowledge, and this generated file in {AIRFLOW_HOME}
+  has been overwritten in the Breeze environment with another CI-specific file. Using ``unit_tests.cfg``
+  as a single source of the configuration for tests - coming from Airflow sources
+  rather than from {AIRFLOW_HOME} is much more convenient and it is automatically used by Pytest.
```
Comment (jedcunningham) [inline TESTING.rst]:
```suggestion
  rather than from {AIRFLOW_HOME} is much more convenient and it is automatically used by pytest.
```

nit

File: airflow/cli/cli_config.py [hunk 12]
```diff
@@ -311,6 +311,26 @@ def string_lower_type(val):
...
+    action="store_true",
+)
+ARG_INCLUDE_EXAMPLES = Arg(
+    ("-e", "--include-examples"), help="Show examples for the configuration variables", action="store_true"
+)
+ARG_INCLUDE_SOURCES = Arg(
+    ("-s", "--include-sources"), help="Show source of the configuration variable", action="store_true"
+)
+ARG_INCLUDE_VARIABLES = Arg(
+    ("-V", "--include-variables"), help="Show environment variable for each option", action="store_true"
```
Comment (jedcunningham) [inline airflow/cli/cli_config.py]: I think the long form should be `--include-env-vars` or `--include-environment-variables`. I don't think we should drop the "env" part though.

File: airflow/config_templates/default_airflow.cfg [hunk 13]
```diff
@@ -16,1507 +15,21 @@
...
-# Path to YAML config. This provides backwards compatibility to pass config as
-# `openlineage.yml` file.
-config_path =
-
-# OpenLineage Client transport configuration. It should contain type
-# and additional options per each type.
+#         airflow config list --defaults
 #
-# Currently supported types are:
+# You can redirect it to your configuration file and edit it:
```
Comment (jedcunningham) [inline airflow/config_templates/default_airflow.cfg:42]: This might be a good place and opportunity to tell people NOT to shove all the defaults into their config file!

File: airflow/configuration.py [hunk 14]
```diff
@@ -1481,77 +1788,24 @@ def initialize_config() -> AirflowConfigParser:
     Called for you automatically as part of the Airflow boot process.
     """
     global FERNET_KEY, AIRFLOW_HOME, WEBSERVER_CONFIG
```
Comment (jedcunningham) [inline airflow/configuration.py]:
```suggestion
    global WEBSERVER_CONFIG
```

These aren't used any longer in this scope.

File: airflow/configuration.py [hunk 15]
```diff
@@ -1260,7 +1483,8 @@ def _deprecated_value_is_set_in_config(
             if config_type == "default":
                 continue
             try:
-                deprecated_section_array = config.items(section=deprecated_section, raw=True)
+                section_items = config.items(section=deprecated_section, raw=True)
+                deprecated_section_array = section_items
```
Comment (jedcunningham) [inline airflow/configuration.py]: Why this change?

File: airflow/cli/cli_config.py (hunk 12 above)
Comment (potiuk) [reply airflow/cli/cli_config.py]: Good point. changing.

File: airflow/config_templates/default_airflow.cfg (hunk 13 above)
Comment (potiuk) [reply airflow/config_templates/default_airflow.cfg:42]: Good point. At least start edifucating people here is a good idea.

Comment (potiuk) [reply airflow/config_templates/default_airflow.cfg:42]: I actually did even more than that. The "--defaults" option now containts all the variables commented out by .... default :) . This is really nice configuration that is "starting point" - because you can still see all the generated descriptions, examples, variables, but the actual values are all commented out, which means that you can selectively uncomment them (and change). I'v also copied this description to the user docs "howto/set-config.rst" and explained this is the approach recommended for production.

File: airflow/configuration.py (hunk 14 above)
Comment (potiuk) [reply airflow/configuration.py]: Right.

File: airflow/configuration.py (hunk 15 above)
Comment (potiuk) [reply airflow/configuration.py]: Remnant of  some other approach or debugging. bringing it back to original state

**Comment by potiuk** [comment]
Applied second round of comments from @jedcunningham.

For those who did not try it yet, the nice part of this change (after @jedcunningham comment) is that we can start promoting the "good" pattern of only changing the configuration variables that users actually want to change.

Part of the change is to make the ``airflow config list`` much more versatile and useful. The default output will (as it did) contain all the configuration (including currently set variables) in a very simple plain format without extra comments and boilerplate as it used to:

Screenshot 2023-07-19 at 07 08 39


But this command is now a bit of "swiss-army-knife" of configuration:

Screenshot 2023-07-19 at 07 09 38

Especially the `--defaults` switch is particularly useful to generate default "production" grade configuration. The configuration will have only the built-in airflow defaults (so it does not include the env variables set locally) and it will contains all the descriptions, and generated corresponding variable name for all the configurations, but all the values are commented out by default - so that user can very selectively only uncomment and change those values they want.

We've been promoting this kind of behaviour as a good pattern (only set those values that you want to be changed from defaults) and with such generated config file it is, I think, perfect way of doing it, because the user can see all the descriptions and examples right in the place the values can be edited, but the values are actually commented out, so it requires deliberate effort to change them - and users will easily see which of the values they actually changed.


Screenshot 2023-07-19 at 07 16 09

**Comment by potiuk** [comment]
Green now. Looking for some more reviews and finally merging to continue with config move to providers and executors move :)

File: airflow/config_templates/default_airflow.cfg (hunk 13 above)
Comment (jedcunningham) [reply airflow/config_templates/default_airflow.cfg:42]: Perfect, thanks! I missed that detail...

Comment (potiuk) [reply airflow/config_templates/default_airflow.cfg:42]: It was not like that before. I added it after your comment when I realized we can finally have a tool to help users do what we always said is a good idea,  so credits go to you for putting me on the path :)

File: airflow/config_templates/default_airflow.cfg [hunk 16]
```diff
@@ -16,1507 +15,31 @@
...
-# Logging level.
+#         airflow config list --defaults
 #
-# Supported values: ``CRITICAL``, ``ERROR``, ``WARNING``, ``INFO``, ``DEBUG``.
-logging_level = INFO
-
-# Logging level for celery. If not set, it uses the value of logging_level
+# This command will produce the output that you can copy to your configuration file and edit.
+# It will contain all the default configuration options nicely commented out, with examples and
+# all the values will be commented out so you can only un-comment and change those that you want to change.
```
Comment (o-nikolas) [inline airflow/config_templates/default_airflow.cfg]: Simplified and removed some duplication
```suggestion
# It will contain all the default configuration options, with examples, nicely commented out
# so you need only un-comment and modify those that you want to change.
```

File: airflow/config_templates/default_airflow.cfg [hunk 17]
```diff
@@ -16,1507 +15,31 @@
...
+# This way you can easily keep track of all the configuration options that you changed from default
+# and you can also easily upgrade your installation to new versions of Airflow when they come out and
+# Automatically use the defaults for existing options if they changed there.
```
Comment (o-nikolas) [inline airflow/config_templates/default_airflow.cfg]: Small typo fix. Also this is **such an amazing sideeffect of this approach!** :rocket: 
```suggestion
# automatically use the defaults for existing options if they changed there.
```

File: docs/apache-airflow/howto/set-config.rst [hunk 18]
```diff
@@ -21,8 +21,31 @@ Setting Configuration Options
...
+However, for production case you are advised to generate the configuration using command line:
+
+.. code-block:: bash
+
+    airflow config list --defaults
+
+This command will produce the output that you can copy to your configuration file and edit.
+
+It will contain all the default configuration options nicely commented out, with examples and
+all the values will be commented out so you can only un-comment and change those that you want to change.
```
Comment (o-nikolas) [inline docs/apache-airflow/howto/set-config.rst]: Same as above:
```suggestion
 It will contain all the default configuration options, with examples, nicely commented out
so you need only un-comment and modify those that you want to change.
```

File: docs/apache-airflow/howto/set-config.rst [hunk 19]
```diff
@@ -21,8 +21,31 @@ Setting Configuration Options
...
+This way you can easily keep track of all the configuration options that you changed from default
+and you can also easily upgrade your installation to new versions of Airflow when they come out and
+Automatically use the defaults for existing options if they changed there.
```
Comment (o-nikolas) [inline docs/apache-airflow/howto/set-config.rst]: Same as above:
```suggestion
automatically use the defaults for existing options if they changed there.
```

**Review by o-nikolas — COMMENTED** [review]
Looked at most of the PR, except mainly the changes in configuration.py

**Review by jedcunningham — APPROVED** [review]

**Review by hussein-awala — APPROVED** [review]

File: airflow/config_templates/default_airflow.cfg (hunk 17 above)
Comment (potiuk) [reply airflow/config_templates/default_airflow.cfg]: > Also this is such an amazing sideeffect of this approach! 

Indeed. Thanks to @jedcunningham for actually making me realise we can do it :)

**Comment by potiuk** [comment]
Random infra-related stability problems. Merging finally :)
